realm/realm-library/src/main/java/io/realm/internal/Context.java

```diff
@@ -17,127 +17,24 @@
 package io.realm.internal;
 
 import java.lang.ref.ReferenceQueue;
-import java.util.ArrayList;
-import java.util.List;
 
+// Currently we free native objects in two threads, the SharedGroup is freed in the caller thread, others are freed in
+// PhantomDaemon thread. And the destruction in both threads are locked by the corresponding context.
+// The purpose of locking on Context is:
+// Destruction of SharedGroup (and hence Group and Table) is currently not thread-safe with respect to destruction of
+// other accessors, you have two ensure mutual exclusion. This is also illustrated by the use of locks in the test
+// test_destructor_thread_safety.cpp. Explicit call of SharedGroup::close() or Table::detach() is also not thread-safe
+// with respect to destruction of other accessors.
 public class Context {
+    private final static ReferenceQueue<NativeObject> referenceQueue = new ReferenceQueue<NativeObject>();
+    private final static Thread phantomThread = new Thread(new FinalizerRunnable(referenceQueue));
```

Comment:
minor: scary name 👻 I know it's soon 🎃 but maybe `gcThread` is more appropriate?